internal/escape/escape.go

```diff
@@ -15,105 +15,122 @@
 // Package escape includes helpers for escaping and unescaping strings.
 package escape
 
-// Escape returns s, with all bytes for which shouldEscape returns true
-// escaped to "__0xXX__", where XX is the hex representation of the byte
+import (
+	"fmt"
+	"strconv"
+)
+
+// Escape returns s, with all runes for which shouldEscape returns true
+// escaped to "__0xXXX__", where XXX is the hex representation of the rune
 // value.
 //
-// shouldEscape takes the whole string and an index instead of a single
-// byte because some escape decisions require context. For example, we
-// might want to escape the second "/" in "//" but not the first one.
-func Escape(shouldEscape func(s string, i int) bool, s string) string {
-	// Do a first pass to see how many bytes need escaping.
+// Note: shouldEscape takes the whole string as a slice of runes and an
+// index. Passing it a single byte or a single rune doesn't provide
+// enough context for some escape decisions; for example, the caller might
+// want to escape the second "/" in "//" but not the first one.
+// We pass a slice of runes instead of the string or a slice of bytes
+// because some decisions will be made on a rune basis (e.g., encode
+// all non-ASCII runes).
+func Escape(s string, shouldEscape func(s []rune, i int) bool) string {
+	// Do a first pass to see which runes (if any) need escaping.
+	runes := []rune(s)
```

Comment:
OK, so now your problem is a string that is not UTF-8. (Maybe some providers support such strings.) In that case I think you'll get a Unicode replacement character instead of the true bytes.

Maybe it's not worth worrying about that now.

Reply:
I added a note that non-UTF8 isn't supported (i.e., non-UTF8 characters are escaped to `unicode.ReplacementChar`). We can always add it later.

When I'm using this package, I'll try non-UTF8 with a few providers to see what happens. If anyone supports it, we should probably support it.